File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> stack;
        unordered_map<int, int> hashmap;
        vector<int> ans;

        for (int num : nums2) {
            while (stack.empty() == false && num > stack.top()) {
                hashmap[stack.top()] = num;
                stack.pop();
            }
            stack.push(num);
        }

        while (stack.empty() == false) {
            hashmap[stack.top()] = -1;
            stack.pop();
        }

        for (int num : nums1) {
            ans.push_back(hashmap[num]);
        }

        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack, ans = [], []
        hash = {}
        for num in nums2:
            while stack and num > stack[-1]:
                hash[stack[-1]] = num
                stack.pop()
            stack.append(num)
        while stack:
            hash[stack[-1]] = -1
            stack.pop()
        for num in nums1:
            ans.append(hash[num])
        return ans
        